src/api/mod.rs

```diff
@@ -3281,4 +3295,94 @@ mod test {
       ]
     );
   }
+
+  #[interpolate_test(8, 8)]
+  #[interpolate_test(10, 10)]
+  #[interpolate_test(16, 16)]
+  fn lookahead_size_properly_bounded(rdo_lookahead: u64) {
+    // Test that lookahead reads in the proper number of frames at once
+
+    let mut ctx = setup_encoder::<u8>(
+      64,
+      80,
+      10,
+      100,
+      8,
+      ChromaSampling::Cs420,
+      0,
+      100,
+      0,
+      false,
+      true,
+      rdo_lookahead
+    );
+
+    let limit = 60;
+
+    let scenechange_lookahead_len = ctx.inner.inter_cfg.keyframe_lookahead_distance() as usize;
+    for frameno in 0..scenechange_lookahead_len {
+      let input = ctx.new_frame();
+      let _ = ctx.send_frame(input);
+      let res = ctx.receive_packet();
+      assert!(res.is_err());
+      assert_eq!(ctx.inner.frame_q.len(), frameno + 1);
+      assert_eq!(ctx.inner.frame_invariants.len(), 0);
+    }
+
+    let rdo_lookahead_len = ctx.config.rdo_lookahead_frames as usize;
+    for frameno in scenechange_lookahead_len..rdo_lookahead_len {
+      let input = ctx.new_frame();
+      let _ = ctx.send_frame(input);
+      let res = ctx.receive_packet();
+      assert!(res.is_err());
+      assert_eq!(ctx.inner.frame_q.len(), frameno + 1);
+      // The exact logic gets complicated here because of frame pyramids,
+      // but the size of the FI queue should always hold to this rule.
+      assert!(ctx.inner.frame_invariants.len() < rdo_lookahead_len);
```

Comment:
Perhaps an array like `let frame_invariants_len_at_step = [...];` that you compare with every step of this loop would be better?

Reply:
Yeah, we could use that to make the whole test simpler as well. Removes the need for multiple loops. I'll do that.